File: markdown-parser/python/tests/document_builder.py
from __future__ import annotations


class DocumentBuilder:
    def __init__(self) -> None:
        self._lines: list[str] = []

    def with_line(self, line: str) -> DocumentBuilder:
        self._lines.append(line)
        return self

    def with_blank_line(self) -> DocumentBuilder:
        self._lines.append("")
        return self

    def with_heading(self, level: int, text: str) -> DocumentBuilder:
        self._lines.append(f"{'#' * level} {text}")
        return self

    def with_paragraph(self, text: str) -> DocumentBuilder:
        self._lines.append(text)
        return self

    def with_list_item(self, text: str) -> DocumentBuilder:
        self._lines.append(f"- {text}")
        return self

    def with_blockquote(self, text: str) -> DocumentBuilder:
        self._lines.append(f"> {text}")
        return self

    def with_code_block(self, content: str) -> DocumentBuilder:
        self._lines.append("```")
        self._lines.append(content)
        self._lines.append("```")
        return self

    def build(self) -> str:
        return "\n".join(self._lines)

File: markdown-parser/typescript/tests/DocumentBuilder.ts
export class DocumentBuilder {
  private lines: string[] = [];

  withLine(line: string): this { this.lines.push(line); return this; }
  withBlankLine(): this { this.lines.push(''); return this; }
  withHeading(level: number, text: string): this { this.lines.push(`${'#'.repeat(level)} ${text}`); return this; }
  withParagraph(text: string): this { this.lines.push(text); return this; }
  withListItem(text: string): this { this.lines.push(`- ${text}`); return this; }
  withBlockquote(text: string): this { this.lines.push(`> ${text}`); return this; }
  withCodeBlock(content: string): this {
    this.lines.push('```');
    this.lines.push(content);
    this.lines.push('```');
    return this;
  }

  build(): string { return this.lines.join('\n'); }
}
